README.md
[![Travis-CI Build Status](https://travis-ci.org/julian-urbano/gt4ireval.svg?branch=master)](https://travis-ci.org/julian-urbano/gt4ireval)
[![CRAN version](http://www.r-pkg.org/badges/version/gt4ireval)](https://cran.r-project.org/package=gt4ireval) 
[![License](https://img.shields.io/badge/license-MIT-blue.svg)](https://opensource.org/licenses/MIT)

# gt4ireval

Provides tools to measure the reliability of an Information Retrieval test collection. It allows users to estimate reliability using Generalizability Theory and map those estimates onto well-known indicators such as Kendall tau correlation or sensitivity.

For a full background please refer to *Julián Urbano, Mónica Marrero and Diego Martín, "[On the Measurement of Test Collection Reliability](http://julian-urbano.info/wp-content/uploads/055-measurement-test-collection-reliability.pdf)", ACM SIGIR, 2013*.

## Installation

You may install the stable release from CRAN

```{r}
install.packages("gt4ireval")
```

or the latest development version from GitHub

``` r
devtools::install_github("julian-urbano/gt4ireval")
```

## Usage

A full user manual in available in the [package vignette](https://cran.r-project.org/web/packages/gt4ireval/vignettes/gt4ireval.html).

As a very simple example, we can analyze the TREC-3 Ad hoc data:

    > head(adhoc3)
        sys1   sys2   sys3   sys4   sys5   sys6   sys7 ...
    1 0.2830 0.5163 0.4810 0.5737 0.5184 0.4945 0.5013 ...
    2 0.0168 0.5442 0.3987 0.2964 0.6115 0.2354 0.1689 ...
    ...

We first run a G-study,

    > ah3.g <- gstudy(adhoc3, drop = 0.25)
    > ah3.g
    
    Summary of G-Study
    
                     Systems     Queries Interaction
                 ----------- ----------- -----------
    Variance       0.0028117    0.028093    0.010152
    Variance(%)       6.8482      68.425      24.727
    ---
    Mean Sq.         0.15074     0.85296    0.010152
    Sample size           30          50        1500

then a D-study,

    > dstudy(ah3.g, queries = c(50, 100, 150))
    
    Summary of D-Study
    
    Call:
        queries = 50 
      stability = 0.95 
          alpha = 0.025 
    
    Stability:
                                               Erho2                                   Phi
                 -----------------------------------   -----------------------------------
         Queries    Expected       Lower       Upper      Expected       Lower       Upper
     ----------- ----------- ----------- -----------   ----------- ----------- -----------
              50     0.93265     0.89311     0.96287       0.78613     0.66141     0.88039 
             100     0.96515     0.94354     0.98109       0.88026     0.79621     0.93639 
             150     0.97649     0.96164     0.98731       0.91686     0.85423     0.95668 
    
    Required number of queries:
                                               Erho2                                   Phi
                 -----------------------------------   -----------------------------------
       Stability    Expected       Lower       Upper      Expected       Lower       Upper
     ----------- ----------- ----------- -----------   ----------- ----------- -----------
            0.95          69          37         114           259         130         487

and possibly map onto AP correlation, for instance,

    > gt2tauAP(Erho2 = c(0.93, 0.95, 0.98))
    [1] 0.7487836 0.8150692 0.9226192

## License

`gt4ireval` is released under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Acknowledgements

> This work is supported by an A4U postdoctoral grant and a Juan de la Cierva postdoctoral fellowship.


CHANGELOG
* Confidence intervals in G-study (#5).
* Significance levels must be in the range (0, 0.5) (#7).
* Allow negative variance estimates (#6).

## gt4ireval 2.0, 06-Mar-2017

* Now as fully fledged R package in CRAN (#4).
* Switched to MIT license (#3).
* Refactored function names and added error checking.
* Separated GT-data mapping into several functions.
* Added simple unit tests from Brennan's book (#2).
* g.study accepts both data.frame and matrix (#1).

## gt4ireval 1.1, 08-May-2014

* Fixed minor bug in print.dstudy showing wrong number of queries in call from g.study.

## gt4ireval 1.0, 09-Dec-2013

* Initial release.


gt4ireval.test.R
## Copyright (C) 2015  Julián Urbano <urbano.julian@gmail.com>
##
## This program is free software: you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation, either version 3 of the License, or
## (at your option) any later version.
##
## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with this program.  If not, see http://www.gnu.org/licenses/.

## Unit tests based on example from:
##   R. L. Brennan, "Generalizability Theory". Springer, 2001.

rm(list = ls(all = T))
library(RUnit)

## Synthetic dataset no. 4 from Table 3.2 on page 73, recasted as a p X i design, as required on page 182.
df <- structure(list(p1 = c(5L, 6L, 5L, 5L, 5L, 3L, 4L, 5L, 6L, 7L, 3L, 3L),
                     p2 = c(9L, 3L, 7L, 7L, 7L, 5L, 5L, 5L, 7L, 7L, 5L, 2L),
                     p3 = c(3L, 4L, 3L, 3L, 5L, 3L, 3L, 5L, 6L, 5L, 1L, 6L),
                     p4 = c(7L, 5L, 5L, 3L, 3L, 1L, 4L, 3L, 5L, 3L, 3L, 5L),
                     p5 = c(9L, 2L, 9L, 7L, 7L, 7L, 3L, 7L, 2L, 7L, 5L, 3L),
                     p6 = c(3L, 4L, 3L, 5L, 3L, 3L, 6L, 3L, 4L, 5L, 1L, 2L),
                     p7 = c(7L, 3L, 7L, 7L, 7L, 5L, 5L, 7L, 5L, 5L, 5L, 4L),
                     p8 = c(5L, 8L, 5L, 7L, 7L, 5L, 5L, 4L, 3L, 2L, 1L, 1L),
                     p9 = c(9L, 9L, 8L, 8L, 6L, 6L, 6L, 5L, 5L, 8L, 1L, 1L),
                     p10 = c(4L, 4L, 4L, 3L, 3L, 5L, 6L, 5L, 5L, 7L, 1L, 1L)),
                .Names = c("p1", "p2", "p3", "p4", "p5", "p6", "p7", "p8", "p9", "p10"),
                class = "data.frame", row.names = c(NA, -12L))

source("gt4ireval.R")

## Test G study with Table 6.2 on page 182
test.g.study <- function() {
  g <- g.study(data = df, drop = 0)

  # Degrees of freedom
  checkEqualsNumeric(10, g$n.s)
  checkEqualsNumeric(12, g$n.q)

  # Variance components
  checkEqualsNumeric(.6258, g$var.s, tolerance = 1e-4)
  checkEqualsNumeric(.8840, g$var.q, tolerance = 1e-4)
  checkEqualsNumeric(2.7872, g$var.e, tolerance = 1e-4)

  # Mean squares
  checkEqualsNumeric(10.2963, g$em.s, tolerance = 1e-4)
  checkEqualsNumeric(11.6273, g$em.q, tolerance = 1e-4)
  checkEqualsNumeric(2.7872, g$em.e, tolerance = 1e-4)
}

## Test D study with Table 6.5 on page 198
test.d.study <- function() {
  g <- g.study(data = df, drop = 0)
  d <- d.study(gdata = g, alpha = .1) # 80% confidence intervals

  # Erho2
  checkEqualsNumeric(.729, d$Erho2, tolerance = 1e-3)
  checkEqualsNumeric(.541, d$Erho2.lwr, tolerance = 1e-3)
  checkEqualsNumeric(.876, d$Erho2.upr, tolerance = 1e-3)
  # Phi
  checkEqualsNumeric(.672, d$Phi, tolerance = 1e-3)
  checkEqualsNumeric(.457, d$Phi.lwr, tolerance = 1e-3)
  checkEqualsNumeric(.845, d$Phi.upr, tolerance = 1e-3)
}

# runTestFile("gt4ireval.test.R")


tests/testthat.R
library(testthat)
library(gt4ireval)

test_check("gt4ireval")


tests/testthat/test.dstudy.R
context("D-study")

#' Test D study with Table 6.5 on page 198.
test_that("Synthetic dataset, valid args", {
  g <- gstudy(synthetic4)
  d <- dstudy(g, alpha = .1) # 80% 2-tailed confidence interval

  expect_equal(d$call$queries, 12)

  expect_equal(d$Erho2, .729, tolerance = 1e-3)
  expect_equal(d$Erho2.lwr, .541, tolerance = 1e-3)
  expect_equal(d$Erho2.upr, .876, tolerance = 1e-3)

  expect_equal(d$Phi, .672, tolerance = 1e-3)
  expect_equal(d$Phi.lwr, .457, tolerance = 1e-3)
  expect_equal(d$Phi.upr, .845, tolerance = 1e-3)
})

test_that("Synthetic dataset, invalid args", {
  g <- gstudy(synthetic4)

  expect_error(dstudy(1:3))
  expect_error(dstudy(g, queries = -1))
  expect_error(dstudy(g, stability = c(.1, -.1)))
  expect_error(dstudy(g, stability = c(.9, 1)))
  expect_error(dstudy(g, alpha = c(-.01, .01)))
  expect_error(dstudy(g, alpha = c(.01, .55)))
})


R/gstudy.R
#' G-study (Generalizability)
#'
#' \code{gstudy} runs a G-study with the given data, assuming a fully crossed design (all systems
#' evaluated on the same queries). It can be used to estimate variance components, possibly with
#' confidence intervals, which can further be used to run a D-study with \code{\link{dstudy}}.
#' Confidence intervals are computed with the Jacknife procedure described in sections 6.1.2 and
#' 6.2.4 of Brennan (2001).
#'
#' @param data A data frame or matrix with the existing effectiveness scores. Systems are columns
#'   and queries are rows.
#' @param ensurePositive Whether to set negative variance estimates to zero or not. Defaults to
#'   \code{FALSE}.
#' @param alpha A vector of confidence levels to compute intervals for the variance components. This
#'   is the probability on each side of the interval, so for a 90\% confidence interval one must set
#'   \code{alpha} to 0.05. When set to \code{NULL}, no intervals are computed.
#' @param drop The fraction of worst-performing systems to drop from the data before analysis.
#'   Defaults to 0 (include all systems).
#'
#' @return An object of class \code{gstudy}, with the following components: \tabular{ll}{
#'   \code{n.s}, \code{n.q} \tab Number of systems and number of queries of the existing data. \cr
#'   \code{var.s}, \code{var.q}, \code{var.e} \tab Variance of the system, query, and residual
#'   effects. \cr \code{em.s}, \code{em.q}, \code{em.e} \tab Mean squares of the system, query and
#'   residual components. \cr \code{call} \tab A list with the existing \code{data} and the
#'   percentage of systems to \code{drop}.\cr }
#'
#' @seealso \code{\link{dstudy}}
#' @author Julián Urbano
#' @references R.L. Brennan (2001). Generalizability Theory. Springer.
#'
#'   L.J. Cronbach, G.C. Gleser, H. Nanda and N. Rajaratnam (1972). The dependability of behavioral
#'   measurements: Theory of generalizability for scores and profiles. Wiley. (pp. 54-57, 66,
#'   70-72).
#'
#'   J. Urbano, M. Marrero and D. Martín (2013). On the Measurement of Test Collection Reliability.
#'   ACM SIGIR, pp. 393-402.
#'
#' @examples
#' g <- gstudy(adhoc3)
#'
#' # same, but drop the 20% worst systems
#' g20 <- gstudy(adhoc3, drop = 0.2)
#'
#' # compute 90% and 95% confidence intervals
#' g <- gstudy(adhoc3, alpha = c(.05, 0.025))
#' @export
#' @importFrom stats quantile
gstudy <- function(data, ensurePositive = TRUE, alpha = 0.025, drop = 0) {
  if(drop < 0 || drop >= 1)
    stop("Fraction of worst systems to drop must be in [0,1)")
  if(!is.data.frame(data) && !is.matrix(data))
    stop("Data must be a data.frame or matrix (columns for systems and rows for queries)")
  if(!is.null(alpha) && any(alpha <= 0 | alpha >= 0.5))
    stop("Significance level must be in (0,0.5)")

  # drop bottom results as indicated
  data <- data[, colMeans(data) >= quantile(colMeans(data), drop)]

  # sizes and means
  n.s <- ncol(data)
  means.s <- colMeans(data)
  n.q <- nrow(data)
  means.q <- rowMeans(data)
  means.all <- mean(means.s)

  # sums of squares and mean squares
  ss.s <- sum((means.s - means.all)^2) * n.q
  ss.q <- sum((means.q - means.all)^2) * n.s
  ss.e <- sum((data - means.all)^2) - ss.s - ss.q
  em.s <- ss.s / (n.s - 1)
  em.q <- ss.q / (n.q - 1)
  em.e <- ss.e / (n.s - 1) / (n.q - 1)

  # variance components
  var.s <- (em.s - em.e) / n.q
  var.q <- (em.q - em.e) / n.s
  var.e <- em.e
  if(var.s < 0)
    warning("Estimate of system variance is negative")
  if(var.q < 0)
    warning("Estimate of query variance is negative")
  if(var.e < 0)
    warning("Estimate of interaction variance is negative")

  # interval estimates: 6.1.2 and 6.2.4 in Brennan 2001.
  if(!is.null(alpha)) {
    theta.s <- matrix(NA, ncol = n.s, nrow = n.q)
    theta.q <- matrix(NA, ncol = n.s, nrow = n.q)
    theta.e <- matrix(NA, ncol = n.s, nrow = n.q)

    suppressWarnings(theta.00 <- gstudy(data, ensurePositive = FALSE, alpha = NULL))
    for(s in 1:n.s) {
      suppressWarnings(theta.s0 <- gstudy(data[,-s], ensurePositive = FALSE, alpha = NULL))
      for(q in 1:n.q) {
        suppressWarnings(theta.0q <- gstudy(data[-q,], ensurePositive = FALSE, alpha = NULL))
        suppressWarnings(theta.sq <- gstudy(data[-q, -s], ensurePositive = FALSE, alpha = NULL))

        theta.s[q, s] <- n.s * n.q * theta.00$var.s - (n.s - 1) * n.q * theta.s0$var.s -
          n.s * (n.q - 1) * theta.0q$var.s + (n.s - 1) * (n.q - 1) * theta.sq$var.s
        theta.q[q, s] <- n.s * n.q * theta.00$var.q - (n.s - 1) * n.q * theta.s0$var.q -
          n.s * (n.q - 1) * theta.0q$var.q + (n.s - 1) * (n.q - 1) * theta.sq$var.q
        theta.e[q, s] <- n.s * n.q * theta.00$var.e - (n.s - 1) * n.q * theta.s0$var.e -
          n.s * (n.q - 1) * theta.0q$var.e + (n.s - 1) * (n.q - 1) * theta.sq$var.e
      }
    }
    suppressWarnings(g <- gstudy(theta.s, drop = 0, ensurePositive = FALSE, alpha = NULL))
    se <- sqrt(g$var.s / n.s + g$var.q / n.q + g$var.e / n.s / n.q)
    var.s.lwr <- var.s - qt(1 - alpha, df = n.s * n.q - 1) * se
    var.s.upr <- var.s + qt(1 - alpha, df = n.s * n.q - 1) * se
    suppressWarnings(g <- gstudy(theta.q, drop = 0, ensurePositive = FALSE, alpha = NULL))
    se <- sqrt(g$var.s / n.s + g$var.q / n.q + g$var.e / n.s / n.q)
    var.q.lwr <- var.q - qt(1 - alpha, df = n.s * n.q - 1) * se
    var.q.upr <- var.q + qt(1 - alpha, df = n.s * n.q - 1) * se
    suppressWarnings(g <- gstudy(theta.e, drop = 0, ensurePositive = FALSE, alpha = NULL))
    se <- sqrt(g$var.s / n.s + g$var.q / n.q + g$var.e / n.s / n.q)
    var.e.lwr <- var.e - qt(1 - alpha, df = n.s * n.q - 1) * se
    var.e.upr <- var.e + qt(1 - alpha, df = n.s * n.q - 1) * se
  }

  if(ensurePositive) {
    var.s <- max(0, var.s)
    var.q <- max(0, var.q)
    var.e <- max(0, var.e)
    if(!is.null(alpha)) {
      var.s.lwr <- max(0, var.s.lwr)
      var.s.upr <- max(0, var.s.upr)
      var.q.lwr <- max(0, var.q.lwr)
      var.q.upr <- max(0, var.q.upr)
      var.e.lwr <- max(0, var.e.lwr)
      var.e.upr <- max(0, var.e.upr)
    }
  }

  res <- list(
    n.s = n.s, n.q = n.q,
    var.s = var.s, var.q = var.q, var.e = var.e,
    em.s = em.s, em.q = em.q, em.e = em.e,
    call = list(data = data, drop = drop, ensurePositive = ensurePositive)
  )
  if(!is.null(alpha)) {
    res$var.s.lwr = var.s.lwr
    res$var.s.upr = var.s.upr
    res$var.q.lwr = var.q.lwr
    res$var.q.upr = var.q.upr
    res$var.e.lwr = var.e.lwr
    res$var.e.upr = var.e.upr
  }
  class(res) <- "gstudy"
  return(res)
}

#' @export
print.gstudy <- function(x, ...) {
  cat("\nSummary of G-Study\n")
  cat("\n                 Systems     Queries Interaction\n")
  cat("             ----------- ----------- -----------\n")
  cat(sep = "", "Variance    ",
      sprintf("%*.5g", 12, x$var.s),
      sprintf("%*.5g", 12, x$var.q),
      sprintf("%*.5g", 12, x$var.e), "\n")
  varsum <- (x$var.s + x$var.q + x$var.e) / 100
  cat(sep = "", "Variance(%) ",
      sprintf("%*.5g", 12, x$var.s / varsum),
      sprintf("%*.5g", 12, x$var.q / varsum),
      sprintf("%*.5g", 12, x$var.e / varsum), "\n")
  cat("---\n")
  cat(sep = "", "Mean Sq.    ",
      sprintf("%*.5g", 12, x$em.s),
      sprintf("%*.5g", 12, x$em.q),
      sprintf("%*.5g", 12, x$em.e), "\n")
  cat(sep = "", "Sample size ",
      sprintf("%*.5g", 12, x$n.s),
      sprintf("%*.5g", 12, x$n.q),
      sprintf("%*.5g", 12, x$n.s * x$n.q), "\n\n")
}


R/gt2tau.R
#' Map GT-based Indicators onto Data-based Indicators
#'
#' Maps Erho2 and Phi scores from Generalizability Theory onto traditional data-based scores like
#' the Kendall tau correlation, AP correlation, power, minor conflict rate and major conflict rate
#' with 2-tailed t-tests, absolute and relative sensitivity, and rooted mean squared error.
#'
#' Take these mappings with a grain of salt. See figure 3 in (Urbano, 20013).
#'
#' @param Erho2 Vector of generalizability coefficients to map from.
#' @param Phi Vector of indices of dependability to map from.
#'
#' @return A vector of data-based indicator values.
#' @seealso \code{\link{dstudy}}
#' @author Julián Urbano
#' @references J. Urbano, M. Marrero and D. Martín (2013). On the Measurement of Test Collection
#'   Reliability. ACM SIGIR, pp. 393-402.
#'
#' @examples
#' g <- gstudy(adhoc3)
#' d <- dstudy(g)
#' gt2tau(d$Erho2)
#' gt2rmse(d$Phi)
#'
#' @export
gt2tau <- function(Erho2) {
  if(any(Erho2 <= 0 | Erho2 >= 1))
    stop("Erho2 must be in (0,1]")
  return(Erho2 ^ 2.84729794002905)
}
#' @rdname gt2tau
#' @export
gt2tauAP <- function(Erho2) {
  if(any(Erho2 <= 0 | Erho2 >= 1))
    stop("Erho2 must be in (0,1]")
  return(Erho2 ^ 3.98652984123827)
}
#' @rdname gt2tau
#' @export
gt2power <- function(Erho2) {
  if(any(Erho2 <= 0 | Erho2 >= 1))
    stop("Erho2 must be in (0,1]")
  return(Erho2 ^ 4.77902509574171)
}
#' @rdname gt2tau
#' @export
gt2minor <- function(Erho2) {
  if(any(Erho2 <= 0 | Erho2 >= 1))
    stop("Erho2 must be in (0,1]")
  return((1-Erho2) ^ 1.53337366741287)
}
#' @rdname gt2tau
#' @export
gt2major <- function(Erho2) {
  if(any(Erho2 <= 0 | Erho2 >= 1))
    stop("Erho2 must be in (0,1]")
  return((1-Erho2) ^ 2.62976839002005)
}
#' @rdname gt2tau
#' @export
gt2asens <- function(Erho2) {
  if(any(Erho2 <= 0 | Erho2 >= 1))
    stop("Erho2 must be in (0,1]")
  return((1-Erho2) ^ 1.54402996734738)
}
#' @rdname gt2tau
#' @export
gt2rsens <- function(Phi) {
  if(any(Phi <= 0 | Phi >= 1))
    stop("Phi must be in (0,1]")
  return((1-Phi) ^ 1.29759126030214)
}
#' @rdname gt2tau
#' @export
gt2rmse <- function(Phi) {
  if(any(Phi <= 0 | Phi >= 1))
    stop("Phi must be in (0,1]")
  return((1-Phi) ^ 3.27642726002903)
}


R/gt4ireval.R
## Copyright (C) 2013-2015  Julián Urbano <urbano.julian@gmail.com>
##
## This program is free software: you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation, either version 3 of the License, or
## (at your option) any later version.
##
## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with this program.  If not, see http://www.gnu.org/licenses/.

g.study <- function(data, drop = 0) {
  if(!is.data.frame(data) & !is.matrix(data))
    stop("Data must be a data.frame or matrix (columns for systems and rows for queries)")

	# drop bottom results as indicated
	t <- data[,colMeans(data) >= quantile(colMeans(data), drop)]

	# sizes and means
	n.s <- ncol(t)
	means.s <- colMeans(t)
	n.q <- nrow(t)
	means.q <- rowMeans(t)
	means.all <- mean(means.s)

	# sums of squares and mean squares
	ss.s <- sum((means.s - means.all)^2) * n.q
	ss.q <- sum((means.q - means.all)^2) * n.s
	ss.e <- sum((t - means.all)^2) - ss.s - ss.q
	em.s <- ss.s / (n.s - 1)
	em.q <- ss.q / (n.q - 1)
	em.e <- ss.e / (n.s - 1) / (n.q - 1)

	# variance components
	var.s <- (em.s - em.e) / n.q
	var.q <- (em.q - em.e) / n.s
	var.e <- em.e
	if(var.s < 0){
		var.s <- 0
		warning("Estimate of system variance is negative")
	}
	if(var.q < 0){
		var.q <- 0
		warning("Estimate of query variance is negative")
	}
	if(var.e < 0){
		var.e <- 0
		warning("Estimate of interaction variance is negative")
	}

	res <- list(
		n.s = n.s, n.q = n.q,
		var.s = var.s, var.q = var.q, var.e = var.e,
		em.s = em.s, em.q = em.q, em.e = em.e
	)
	class(res) <- "gstudy"
	return(res)
}

d.study <- function(gdata, queries = gdata$n.q, stability = 0.95, alpha = 0.025) {
	if(length(alpha) > 1 & length(queries) > 1)
		stop("Only one of 'alpha' and 'queries' may have multiple values")
	if(length(alpha) > 1 & length(stability) > 1)
		stop("Only one of 'alpha' and 'stability' may have multiple values")

	# Point estimates for the indicated n.q
	Erho2_ <- gdata$var.s / (gdata$var.s + gdata$var.e / queries)
	n.q_Erho2 <- ceiling((stability * gdata$var.e) / (gdata$var.s * (1 - stability)))

	Phi_ <- gdata$var.s / (gdata$var.s + (gdata$var.q + gdata$var.e) / queries)
	n.q_Phi <- ceiling((stability * (gdata$var.q + gdata$var.e)) / (gdata$var.s * (1 - stability)))

	# Interval estimates
	df.s <- gdata$n.s - 1
	df.q <- gdata$n.q - 1
	df.e <- df.s * df.q

	Erho2.lwr <- gdata$em.s / (gdata$em.e * qf(1-alpha, df.s, df.e))
	Erho2.lwr <- (Erho2.lwr - 1) / gdata$n.q
	n.q_Erho2.lwr <- ceiling(stability / (Erho2.lwr * (1 - stability)))
	Erho2.lwr <- queries * Erho2.lwr / (1 + queries * Erho2.lwr)

	Erho2.upr <- gdata$em.s / (gdata$em.e * qf(alpha, df.s, df.e))
	Erho2.upr <- (Erho2.upr - 1) / gdata$n.q
	n.q_Erho2.upr <- ceiling(stability / (Erho2.upr * (1 - stability)))
	Erho2.upr <- queries * Erho2.upr / (1 + queries * Erho2.upr)

	Phi.lwr <- gdata$em.s^2 - qf(1-alpha, df.s, Inf) * gdata$em.s * gdata$em.e
	Phi.lwr <- Phi.lwr + (qf(1-alpha, df.s, Inf) - qf(1-alpha, df.s, df.e)) *
		qf(1-alpha, df.s, df.e) * gdata$em.e^2
	Phi.lwr <- Phi.lwr / ( (gdata$n.s - 1) * qf(1-alpha, df.s, Inf,) * gdata$em.s * gdata$em.e +
		qf(1-alpha, df.s, df.q) * gdata$em.s * gdata$em.q )
	Phi.lwr <- gdata$n.s * Phi.lwr / (gdata$n.s * Phi.lwr + gdata$n.q)
	n.q_Phi.lwr <- ceiling(stability * (1 - Phi.lwr) / (Phi.lwr * (1 - stability)))
	Phi.lwr <- queries * Phi.lwr / (1 + (queries - 1) * Phi.lwr)

	Phi.upr <- gdata$em.s^2 - qf(alpha, df.s, Inf) * gdata$em.s * gdata$em.e
	Phi.upr <- Phi.upr + (qf(alpha, df.s, Inf) - qf(alpha, df.s, df.e)) *
		qf(alpha, df.s, df.e) * gdata$em.e^2
	Phi.upr <- Phi.upr / ( (gdata$n.s - 1) * qf(alpha, df.s, Inf,) * gdata$em.s * gdata$em.e +
		qf(alpha, df.s, df.q) * gdata$em.s * gdata$em.q )
	Phi.upr <- gdata$n.s * Phi.upr / (gdata$n.s * Phi.upr + gdata$n.q)
	n.q_Phi.upr <- ceiling(stability * (1 - Phi.upr) / (Phi.upr * (1 - stability)))
	Phi.upr <- queries * Phi.upr / (1 + (queries - 1) * Phi.upr)

	res <- list(
		Erho2 = Erho2_, Phi = Phi_,
		n.q_Erho2 = n.q_Erho2, n.q_Phi = n.q_Phi,
		Erho2.lwr = Erho2.lwr, Erho2.upr = Erho2.upr,
		Phi.lwr = Phi.lwr, Phi.upr = Phi.upr,
		n.q_Erho2.lwr = n.q_Erho2.lwr, n.q_Erho2.upr = n.q_Erho2.upr,
		n.q_Phi.lwr = n.q_Phi.lwr, n.q_Phi.upr = n.q_Phi.upr,
		call = list(gstudy = gdata,
			queries = queries, stability = stability, alpha = alpha
		)
	)
	class(res) <- "dstudy"
	return(res)
}

gt2data <- function(Erho2 = 0.95, Phi = 0.95) {
	as <- c(tau = 2.84729794002905, tau.ap = 3.98652984123827, power = 4.77902509574171, minor = 1.53337366741287, major = 2.62976839002005, asens = 1.54402996734738, rsens = 1.29759126030214, rmse = 3.27642726002903)

	res <- list(
		tau = Erho2^as["tau"], tau.ap = Erho2^as["tau.ap"],
		power = Erho2^as["power"], minor = (1-Erho2)^as["minor"], major = (1-Erho2)^as["major"],
		asens = (1-Erho2)^as["asens"], rsens = (1-Phi)^as["rsens"],
		rmse = (1-Phi)^as["rmse"],
		call = list(Erho2 = Erho2, Phi = Phi)
	)
	class(res) <- "gt2data"
	return(res)
}

print.gstudy <- function(x) {
	cat("\nSummary of G-Study\n")
	cat("\n                 Systems     Queries Interaction\n")
	cat("             ----------- ----------- -----------\n")
	cat(sep="", "Variance    ", sprintf("%*.5g",12,x$var.s), sprintf("%*.5g",12,x$var.q), sprintf("%*.5g",12,x$var.e),"\n")
	varsum <- (x$var.s + x$var.q + x$var.e) / 100
	cat(sep="", "Variance(%) ", sprintf("%*.5g",12,x$var.s/varsum), sprintf("%*.5g",12,x$var.q/varsum), sprintf("%*.5g",12,x$var.e/varsum),"\n")
	cat("---\n")
	cat(sep="", "Mean Sq.    ", sprintf("%*.5g",12,x$em.s), sprintf("%*.5g",12,x$em.q), sprintf("%*.5g",12,x$em.e),"\n")
	cat(sep="", "Sample size ", sprintf("%*.5g",12,x$n.s), sprintf("%*.5g",12,x$n.q), sprintf("%*.5g",12,x$n.s*x$n.q), "\n\n")
}

print.dstudy <- function(x) {
	cat("\nSummary of D-Study\n")
	cat("\nCall:\n")
	cat("    queries =", x$call$queries, "\n")
	cat("  stability =", x$call$stability, "\n")
	cat("      alpha =", x$call$alpha, "\n")

	# Stability
	cat("\nStability:\n")
	if(length(x$call$alpha) > 1){
		cat("                                           Erho2                                   Phi\n")
		cat("             -----------------------------------   -----------------------------------\n")
		cat("       Alpha    Expected       Lower       Upper      Expected       Lower       Upper\n")
		cat(" ----------- ----------- ----------- -----------   ----------- ----------- -----------\n")
		for(i in 1:length(x$call$alpha))
			cat("", sprintf("%*.5g",11,x$call$alpha[i]), sprintf("%*.5g",11,x$Erho2[1]), sprintf("%*.5g",11,x$Erho2.lwr[i]), sprintf("%*.5g",11,x$Erho2.upr[i]), " ", sprintf("%*.5g",11,x$Phi[1]), sprintf("%*.5g",11,x$Phi.lwr[i]), sprintf("%*.5g",11,x$Phi.upr[i]), "\n")
	}else{
		cat("                                           Erho2                                   Phi\n")
		cat("             -----------------------------------   -----------------------------------\n")
		cat("     Queries    Expected       Lower       Upper      Expected       Lower       Upper\n")
		cat(" ----------- ----------- ----------- -----------   ----------- ----------- -----------\n")
		for(i in 1:length(x$call$queries))
			cat("", sprintf("%*.5g",11,x$call$queries[i]), sprintf("%*.5g",11,x$Erho2[i]), sprintf("%*.5g",11,x$Erho2.lwr[i]), sprintf("%*.5g",11,x$Erho2.upr[i]), " ", sprintf("%*.5g",11,x$Phi[i]), sprintf("%*.5g",11,x$Phi.lwr[i]), sprintf("%*.5g",11,x$Phi.upr[i]), "\n")
	}

	# Number of queries
	cat("\nRequired number of queries:\n")
	if(length(x$call$alpha) > 1){
		cat("                                           Erho2                                   Phi\n")
		cat("             -----------------------------------   -----------------------------------\n")
		cat("       Alpha    Expected       Lower       Upper      Expected       Lower       Upper\n")
		cat(" ----------- ----------- ----------- -----------   ----------- ----------- -----------\n")
		for(i in 1:length(x$call$alpha))
			cat("", sprintf("%*.5g",11,x$call$alpha[i]), sprintf("%*.5g",11,x$n.q_Erho2[1]), sprintf("%*.5g",11,x$n.q_Erho2.upr[i]), sprintf("%*.5g",11,x$n.q_Erho2.lwr[i]), " ", sprintf("%*.5g",11,x$n.q_Phi[1]), sprintf("%*.5g",11,x$n.q_Phi.upr[i]), sprintf("%*.5g",11,x$n.q_Phi.lwr[i]), "\n")
	}else{
		cat("                                           Erho2                                   Phi\n")
		cat("             -----------------------------------   -----------------------------------\n")
		cat("   Stability    Expected       Lower       Upper      Expected       Lower       Upper\n")
		cat(" ----------- ----------- ----------- -----------   ----------- ----------- -----------\n")
		for(i in 1:length(x$call$stability))
			cat("", sprintf("%*.5g",11,x$call$stability[i]), sprintf("%*.5g",11,x$n.q_Erho2[i]), sprintf("%*.5g",11,x$n.q_Erho2.upr[i]), sprintf("%*.5g",11,x$n.q_Erho2.lwr[i]), " ", sprintf("%*.5g",11,x$n.q_Phi[i]), sprintf("%*.5g",11,x$n.q_Phi.upr[i]), sprintf("%*.5g",11,x$n.q_Phi.lwr[i]), "\n")
	}
}

print.gt2data <- function(x) {
	cat("\nEstimated indicators\n")
	cat("\n       Erho2         Tau       TauAP       Power Min. Confl. Maj. Confl.  Abs. Sens.\n")
	cat(" ----------- ----------- ----------- ----------- ----------- ----------- -----------\n")
	for(i in 1:length(x$call$Erho2))
		cat(sep=" ", sprintf("%*.5g",12,x$call$Erho2[i]), sprintf("%*.5g",11,x$tau[i]), sprintf("%*.5g",11,x$tau.ap[i]), sprintf("%*.5g",11,x$power[i]), sprintf("%*.5g",11,x$minor[i]), sprintf("%*.5g",11,x$major[i]), sprintf("%*.5g",11,x$asens[i]), "\n")
	cat("\n         Phi  Rel. Sens.       RMSE\n")
	cat(" ----------- ----------- -----------\n")
	for(i in 1:length(x$call$Phi))
		cat(sep=" ", sprintf("%*.5g",12,x$call$Phi[i]), sprintf("%*.5g",11,x$rsens[i]), sprintf("%*.5g",11,x$rmse[i]), "\n")
}

cat("GT4IREval 1.1 Copyright (C) 2013-2015 Julian Urbano <urbano.julian@gmail.com>\n")
cat("This program comes with ABSOLUTELY NO WARRANTY.\n")
cat("This is free software, and you are welcome to redistribute it\n")
cat("under the terms of the GNU General Public License version 3.\n")
cat("-------------------------------------------\nUse functions g.study , d.study and gt2data\n")
